File: apps/api/src/lib/tokens.ts
import jwt from 'jsonwebtoken';
import { v4 as uuidv4 } from 'uuid';
import { env } from '../config/env';

export interface TokenPayload {
  userId: string;
  tenantId: string | null;
  role: 'ADMIN' | 'CLIENT_OWNER' | 'CLIENT_VIEWER';
}

export interface AccessTokenPair {
  accessToken: string;
  refreshToken: string;
  expiresIn: string;
}

export function generateAccessToken(payload: TokenPayload): string {
  return jwt.sign(payload as object, env.JWT_SECRET, {
    expiresIn: env.JWT_EXPIRES_IN,
  } as jwt.SignOptions);
}

export function generateRefreshToken(): string {
  return uuidv4();
}

export function verifyAccessToken(token: string): TokenPayload {
  return jwt.verify(token, env.JWT_SECRET) as TokenPayload;
}

export function getRefreshTokenExpiry(): Date {
  const duration = parseDuration(env.JWT_REFRESH_EXPIRES_IN);
  return new Date(Date.now() + duration);
}

function parseDuration(duration: string): number {
  const match = duration.match(/^(\d+)([smhd])$/);
  if (!match) {
    return 7 * 24 * 60 * 60 * 1000; // Default 7 days
  }
  
  const value = parseInt(match[1], 10);
  const unit = match[2];
  
  switch (unit) {
    case 's': return value * 1000;
    case 'm': return value * 60 * 1000;
    case 'h': return value * 60 * 60 * 1000;
    case 'd': return value * 24 * 60 * 60 * 1000;
    default: return 7 * 24 * 60 * 60 * 1000;
  }
}

export function generateTokenPair(payload: TokenPayload): AccessTokenPair {
  const accessToken = generateAccessToken(payload);
  const refreshToken = generateRefreshToken();
  
  return {
    accessToken,
    refreshToken,
    expiresIn: env.JWT_EXPIRES_IN,
  };
}

File: apps/api/app/auth/tokens.py
import re
import uuid
import jwt
from datetime import datetime, timedelta
from typing import Optional, Dict, Any
from ..config import settings

def parse_duration_seconds(duration: str) -> int:
    match = re.match(r"^(\d+)([smhd])$", duration.strip())
    if not match:
        return 7 * 86400
    val = int(match.group(1))
    unit = match.group(2)
    if unit == "s":
        return val
    elif unit == "m":
        return val * 60
    elif unit == "h":
        return val * 3600
    elif unit == "d":
        return val * 86400
    return 7 * 86400

def generate_access_token(user_id: str, tenant_id: Optional[str], role: str) -> str:
    exp_seconds = parse_duration_seconds(settings.JWT_EXPIRES_IN)
    payload = {
        "userId": str(user_id),
        "tenantId": str(tenant_id) if tenant_id else None,
        "role": role,
        "exp": datetime.utcnow() + timedelta(seconds=exp_seconds),
        "iat": datetime.utcnow()
    }
    return jwt.encode(payload, settings.JWT_SECRET, algorithm="HS256")

def generate_refresh_token() -> str:
    return str(uuid.uuid4())

def get_refresh_token_expiry() -> datetime:
    exp_seconds = parse_duration_seconds(settings.JWT_REFRESH_EXPIRES_IN)
    return datetime.utcnow() + timedelta(seconds=exp_seconds)

def verify_access_token(token: str) -> Dict[str, Any]:
    try:
        return jwt.decode(token, settings.JWT_SECRET, algorithms=["HS256"])
    except jwt.PyJWTError as e:
        raise ValueError(f"Invalid access token: {e}")

def generate_token_pair(user_id: str, tenant_id: Optional[str], role: str) -> Dict[str, Any]:
    access_token = generate_access_token(user_id, tenant_id, role)
    refresh_token = generate_refresh_token()
    return {
        "accessToken": access_token,
        "refreshToken": refresh_token,
        "expiresIn": settings.JWT_EXPIRES_IN
    }